onnxruntime/core/providers/cuda/math/inverse.cc

```diff
@@ -0,0 +1,153 @@
+      if (std::is_same<T, MLFloat16>::value) {
+        // Convert from MLFloat16(half) to float
+        Impl_Cast<CudaT, float>(reinterpret_cast<const CudaT*>(input.Data<MLFloat16>()), input_workspace.get(), input_count);
+      } else {
+        CUDA_RETURN_IF_ERROR(cudaMemcpy(input_workspace.get(), input.Data<float>(), sizeof(float) * input_count,
+                                        cudaMemcpyDeviceToDevice));
+      }
+      IAllocatorUniquePtr<float*> matrix_ptrs = inst->GetScratchBuffer<float*>(n_batches);
+      ORT_RETURN_IF_ERROR(ComputeMatrixOffsets<float>(input_workspace.get(), num_batches, rows, matrix_ptrs));
+      // Do LU factorization
+      CUBLAS_RETURN_IF_ERROR(cublasSgetrfBatched(cublas_h, dim, matrix_ptrs.get(), dim, pivots.get(), info.get(), n_batches));
+      ORT_RETURN_IF_ERROR(CheckForSingularity(info, info_cpu, num_batches));
+
+      // Need to compute ptrs for output buffers
+      // Output for MLFloat
+      IAllocatorUniquePtr<float*> output_ptrs = inst->GetScratchBuffer<float*>(n_batches);
+      if (std::is_same<T, MLFloat16>::value) {
+        IAllocatorUniquePtr<float> ml_float_output = inst->GetScratchBuffer<float>(input_count);
+        ORT_RETURN_IF_ERROR(ComputeMatrixOffsets<float>(ml_float_output.get(), num_batches, rows, output_ptrs));
+        // Do the inverse
+        CUBLAS_RETURN_IF_ERROR(cublasSgetriBatched(cublas_h, dim, matrix_ptrs.get(), dim, pivots.get(), output_ptrs.get(), dim, info.get(), n_batches));
+        ORT_RETURN_IF_ERROR(CheckForSingularity(info, info_cpu, num_batches));
+        // Copy the result to output with casting
+        Impl_Cast<float, CudaT>(ml_float_output.get(), reinterpret_cast<CudaT*>(output.MutableData<MLFloat16>()), input_count);
+        // We are done here
+      } else {
+        ORT_RETURN_IF_ERROR(ComputeMatrixOffsets<float>(output.MutableData<float>(), num_batches, rows, output_ptrs));
+        // Do the inverse
+        CUBLAS_RETURN_IF_ERROR(cublasSgetriBatched(cublas_h, dim, matrix_ptrs.get(), dim, pivots.get(), output_ptrs.get(), dim, info.get(), n_batches));
+        ORT_RETURN_IF_ERROR(CheckForSingularity(info, info_cpu, num_batches));
+        // We are done here
+      }
+    } else if (std::is_same<T, double>::value) {
+      IAllocatorUniquePtr<double> input_workspace = inst->GetScratchBuffer<double>(static_cast<int>(input_count));
+      CUDA_RETURN_IF_ERROR(cudaMemcpy(input_workspace.get(), input.Data<double>(), sizeof(double) * input_count,
+                                      cudaMemcpyDeviceToDevice));
+
+      IAllocatorUniquePtr<double*> matrix_ptrs = inst->GetScratchBuffer<double*>(n_batches);
+      ORT_RETURN_IF_ERROR(ComputeMatrixOffsets<double>(input_workspace.get(), num_batches, rows, matrix_ptrs));
+      // Do LU factorization
+      CUBLAS_RETURN_IF_ERROR(cublasDgetrfBatched(cublas_h, dim, matrix_ptrs.get(), dim, pivots.get(), info.get(), n_batches));
```

Comment:
Why not use this - https://docs.nvidia.com/cuda/cublas/index.html#cublas-lt-t-gt-matinvbatched ? It seems to basically do getrfBatched + getribatched (provided rows and cols are < 32) ?

Reply:
I will check it. Ke suggested to check [https://docs.nvidia.com/cuda/cusolver/#cuds-intro](https://docs.nvidia.com/cuda/cusolver/#cuds-intro)

---
In reply to: [406942817](https://github.com/microsoft/onnxruntime/pull/3485#discussion_r406942817) [](ancestors = 406942817)

Reply:
It looks like to use cuSolver one must generate an I matrix on a device  for every invocation and that requires writing a kernel which we are trying to avoid. Also TF and PyTorch are using cuBlas so this impl should be good enough.

---
In reply to: [406945634](https://github.com/microsoft/onnxruntime/pull/3485#discussion_r406945634) [](ancestors = 406945634,406942817)